# Filter PhyloPyPruner Results
This repository hosts the updated version of the Filter PhyloPyPruner Result, which was initially developed for the Klebsormidiophyceae phylogenomic study. As the tool has grown and gained more features over time, I've decided to maintain each tool for the PhyloRSeq++ pipeline in its own dedicated GitHub repository for clarity and modularity

This tool filters and processes FASTA alignment files generated by [PhyloPyPruner](https://github.com/fethalen/phylopypruner). It checks for taxonomic group thresholds, optionally cleans up headers, and can strip alignment gaps, which are helpful when re-aligning sequences later with different tools.

## Quick Start

To run the filtering step, edit lines 3 to 9 in the script:  
[FilterPPP_Result_AndRemoveHeadersNumbers.sh](https://github.com/mjbieren/FilterPPPResult/blob/main/FilterPPP_Result_AndRemoveHeadersNumbers.sh)

> 💡 This tool is lightweight. It runs on a single CPU, requires <1 GB RAM, and works on any 64-bit Linux system. Execution time is determined mainly by disk/network I/O (~95%).

If you need to change the memory cap or handle large files, feel free to reach out:  
📧 maaikejacobine.bierenbroodspot@uni-goettingen.de

---

## What It Does

- **Parses PhyloPyPruner FASTA files**
- **Filters by taxonomic group threshold**
- **Optionally removes gene IDs (headers only retain strain/species names (`-h` option))**
- **Optionally removes alignment gaps (`-a` option)**

Only FASTA files meeting the minimum number of distinct taxonomic groups are kept and copied to the output directory.

---

## Executables

This tool was created using the Boost library (version 1.88). It is developed in Visual Studio 2019 with the GCC compiler (For remote Linux). I've compiled two different (static) executables (they are portable!) under:

- **Debian 12:**  
  [FilterPPPResult_Static_Debian.out](https://github.com/mjbieren/FilterPPPResult/blob/main/Sources/Executables/FilterPPPResult_Static_Debian.out)

- **Rocky Linux 8 (Red Hat-based):**  
  [FilterPPPResult_Static_HPC.out](https://github.com/mjbieren/FilterPPPResult/blob/main/Sources/Executables/FilterPPPResult_Static_HPC.out)


Or compile your own from source.

---

## Usage
Generally, the program can run with:
```
FilterPPPResult.out -f [FastaInputPath] -r [OutputFolder] -t [TaxGroupFile] -n [Threshold] -s [SummaryPath] [-h] [-a]
```

### Options

| Flag | Description |
|------|-------------|
| `-f` | Path to FASTA files (required) |
| `-r` | Output directory (required) |
| `-t` | Taxonomic group file (required) |
| `-n` | Threshold: min number of taxonomic groups required (required) |
| `-s` | Output path for summary file (optional) |
| `-h` | Retain only strain/species names in headers (optional) |
| `-a` | Remove alignment gaps (`-` characters) (optional) |

### Taxonomic Group File Format
```
Species1,Species2
Species3,Species4
...
```

Each line defines a distinct taxonomic group.

Examples of taxonomic group files can be found under: [TaxonomicGroupFiles](https://github.com/mjbieren/FilterPPPResult/tree/main/TaxonomicGroupFiles).

### 🔍 Recommendation for Final Tree Inference

When performing final tree inference (e.g., creating a supermatrix), it is **recommended** to choose a filtering threshold that results in approximately **2,000 to 3,000 loci**.

- Using a **lower threshold**  reduces bootstrap support across the tree.
- Increasing the threshold **beyond this range** typically does **not improve support values** (which often plateau at 100%) and will **greatly increase computational time and resource usage**.

Keeping the number of loci within this range helps balance **accuracy** and **efficiency**.

## Example Summary Output
```
There were a total of 1981 Fasta Files in the PhyloPyPruner output directory, which was based on 63 Strains.
The Taxonomic Group file contained a total of 40 Taxonomic Groups.
The user defined 3 as a threshold, meaning 3/40 Taxonomic Groups had to be present for each Fasta file.
From the total of 1981 Fasta files, 212 were filtered out, resulting in 1769 files retained.
Each of these filtered Fasta files was written into the output folder.
```
